src/features/datasets/server/datasetRouter.ts

```diff
@@ -0,0 +1,345 @@
+import { z } from "zod";
+
+import {
+  createTRPCRouter,
+  protectedProjectProcedure,
+} from "@/src/server/api/trpc";
+import { type DatasetRuns, Prisma, type Dataset } from "@prisma/client";
+
+export const datasetRouter = createTRPCRouter({
+  allDatasets: protectedProjectProcedure
+    .input(
+      z.object({
+        projectId: z.string(),
+      }),
+    )
+    .query(async ({ input, ctx }) => {
+      return ctx.prisma.$queryRaw<
+        Array<
+          Dataset & {
+            countDatasetItems: number;
+            countDatasetRuns: number;
+            lastRunAt: Date;
+          }
+        >
+      >(Prisma.sql`
+        SELECT
+          d.id,
+          d.name,
+          d.created_at "createdAt",
+          d.updated_at "updatedAt",
+          d.status,
+          count(distinct di.id)::int "countDatasetItems",
+          count(distinct dr.id)::int "countDatasetRuns",
+          max(dr.created_at) "lastRunAt"
+        FROM datasets d
+        LEFT JOIN dataset_items di ON di.dataset_id = d.id
+        LEFT JOIN dataset_runs dr ON dr.dataset_id = d.id
+        WHERE d.project_id = ${input.projectId}
+        GROUP BY 1,2,3,4,5
+        ORDER BY d.status ASC, d.created_at DESC
+      `);
+    }),
+  byId: protectedProjectProcedure
+    .input(
+      z.object({
+        projectId: z.string(),
+        datasetId: z.string(),
+      }),
+    )
+    .query(async ({ input, ctx }) => {
+      return ctx.prisma.dataset.findUnique({
+        where: {
+          id: input.datasetId,
+          projectId: input.projectId,
+        },
+      });
+    }),
+  runsByDatasetId: protectedProjectProcedure
+    .input(
+      z.object({
+        projectId: z.string(),
+        datasetId: z.string(),
+      }),
+    )
+    .query(async ({ input, ctx }) => {
+      const runs = await ctx.prisma.$queryRaw<
+        Array<
+          DatasetRuns & {
+            countRunItems: number;
+            scores: Record<string, number>;
+          }
+        >
+      >(Prisma.sql`
+        WITH avg_scores_by_run_id AS (
+          SELECT
+            ri.dataset_run_id run_id,
+            s.name score_name,
+            AVG(s.value) AS average_score_value
```

Comment:
FYI: https://www.postgresql.org/docs/current/queries-select-lists.html#QUERIES-COLUMN-LABELS

Reply:
Would keep it as it is for speed.